### ISSUE TITLE
Predictions for (regr.)blackboost differ

### ISSUE THREAD

**Issue by MariaErdmann** [mlr-org/mlr#1244]
I just checked regr.blackboost so far, but I think this applies also for the other blackboost learners.
Tests do not fail because they work around the "bug" (they do not test with an empty list of params):
Mini-example:

```
set.seed(760)
d = getTaskData(regr.task)
orig.mod = mboost::blackboost(medv ~ ., data = d)
orig.pred = predict(orig.mod, newdata = d)

lrn = makeLearner("regr.blackboost")
mod = train(lrn, regr.task)
pred = predict(mod, regr.task)

# Predictions are different
all.equal(as.vector(orig.pred), pred$data$response)


orig.mod.b = mboost::blackboost(medv ~ ., data = d, tree_controls = party::ctree_control(maxdepth = 0))
orig.pred.b = predict(orig.mod.b, newdata = d)

# now they are equal
all.equal(as.vector(orig.pred.b), pred$data$response)

```

Mlr's default settings are not equal to blackboost's default settings, because I need to explicitely set the maxdepth argument. Looking at the blackboost code, I thought maxdepth = 2 is the default, but changing the default to 2 in mlr, did not change the fact that predictions are different.

So I have the following questions:
Is the difference that occurs here, really due to different default settings?
If yes, I don't know where to look for the "real" tree_controls default settings.
I have the same problem also with the other control parameters of blackboost, when I implemented more families (also for count data). Then the only way to get equal results (in the tests) is by specifiying all parameters in the par.vals of mlr and in tree_controls, as well as in boost_control.

Does someone have an idea what is wrong here?

**Comment by MariaErdmann** [comment]
Talking to @berndbischl, the reason why this is happening is as follows:

If one calls (normal) blackboost without specifying any parameters then the, parameters for tree_controls are like that:
````
blackboost(formula, data = list(),
           weights = NULL, na.action = na.pass,
           tree_controls = party::ctree_control(
               teststat = "max",
               testtype = "Teststatistic",
               mincriterion = 0,
               maxdepth = 2, savesplitstats = FALSE),
           ...)
````

But these are not the default values of party::ctree_control, which are:
````
ctree_control(teststat = c("quad", "max"), 
              testtype = c("Bonferroni", "MonteCarlo", 
                           "Univariate", "Teststatistic"), 
              mincriterion = 0.95, minsplit = 20, minbucket = 7, 
              stump = FALSE, nresample = 9999, maxsurrogate = 0, 
              mtry = 0, savesplitstats = TRUE, maxdepth = 0)
````

So right now the mlr blackboost learner is trained with the defaults of party::ctree_control and not with the defaults of tree_controls of blackboost.

I will try to solve this problem together with the mboost.family implementation i PR #1023.

**Comment by MariaErdmann** [comment]
I tried to solve this in #1023 but branches got mixed up and I couldn't solve the mix up now. Anyway, the way we wanted to implement it, dind''t work yet when I tried it out. So this issue is unfortunately still open.
So this is what we worked out:

- Change the default parameter values of tree_controls to the defaults. You find them as you see here
````
blackboost(formula, data = list(),
           weights = NULL, na.action = na.pass,
           tree_controls = party::ctree_control(
               teststat = "max",
               testtype = "Teststatistic",
               mincriterion = 0,
               maxdepth = 2, savesplitstats = FALSE),
           ...)
````

- Pass all control parameters (tree_controls and boost_control) to the signature of the train function

- write a function something like
makeBoostControlParameter = function(teststat = "max", testtype = "Teststatistic", mincriterion = 0, maxdepth = 2, savesplitstats, stump, minsplit, minbucket, nresample, maxsurrogate, mtry)

- pass makeBoostControl to learnerArgsToControl

````
ct =  learnerArgsToControl(makeBoostTreeControl)
````

- But still this seems not to be the end of the story to get the same predictions

Sorry, that I couldn't fix it. I am on holiday now, so probably somebody wants to take this over.

**Comment by catastrophic-failure** [comment]
Should `git revert SHA` work in this case?

**Comment by larskotthoff** [comment]
In principle yes, but what you want to do depends on what exactly the mix up is.

**Comment by MariaErdmann** [comment]
I did 'git reset' on my branch mboost.families (see PR #1023 (this was before I posted this)), where I tried to resolve this issue as well, which was probably not the best idea...

**Comment by MariaErdmann** [comment]
Fixed this in #1023. The approach is similar to that one of cforest learner and differs from the approach I primarily explained here.
